File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.py
from collections import deque


def solution(cacheSize, cities):
    answer = 0
    cache = deque()
    
    for time, city in enumerate(cities):
        city = city.lower()
        
        if city in cache:
            answer += 1
            cache.remove(city)
            cache.append(city)
            continue
        
        answer += 5
        cache.append(city)
        if len(cache) > cacheSize:
            cache.popleft()
    
    return answer

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.*;

public class Solution {

    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        
        Deque<String> cache = new LinkedList<>();
        
        if (cacheSize == 0) {
            return cities.length * 5;
        }
        
        for (String city : cities) {
            city = city.toLowerCase();
            
            if (cache.contains(city)) {
                answer += 1;
                cache.remove(city);
                cache.addLast(city);
            } else {
                answer += 5;
                cache.addLast(city);
                
                if (cache.size() > cacheSize) {
                    cache.pollFirst();
                }
            }
        }
        
        return answer;
    }
}